File: lib/ts/src/internal/validation/ValidateStruct.ts
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

import { TFieldDeclaration } from '../types/TFieldDeclaration.js';
import { ValidationFailure } from '../../internal/validation/ValidationFailure.js';
import { getTypeUnexpectedValidationFailure } from '../../internal/validation/GetTypeUnexpectedValidationFailure.js';
import { validateStructFields } from '../../internal/validation/ValidateStructFields.js';
import { structName } from '../types/TStruct.js';
import { ValidateContext } from './ValidateContext.js';

export function validateStruct(
    value: any,
    name: string,
    fields: Record<string, TFieldDeclaration>,
    ctx: ValidateContext,
): ValidationFailure[] {
    if (typeof value === 'object' && !Array.isArray(value)) {
        const selectedFields = ctx.select?.[name] ?? null;
        return validateStructFields(fields, selectedFields, value, ctx);
    } else {
        return getTypeUnexpectedValidationFailure([], value, structName);
    }
}

File: tool/uapicodegen/uapicodegen/uapi/internal/validation/ValidateStruct.py
from typing import TYPE_CHECKING, cast
from ...internal.types.UStruct import _STRUCT_NAME

if TYPE_CHECKING:
    from ...internal.validation.ValidateContext import ValidateContext
    from ...internal.types.UFieldDeclaration import UFieldDeclaration
    from ...internal.types.UTypeDeclaration import UTypeDeclaration
    from ...internal.validation.ValidationFailure import ValidationFailure


def validate_struct(value: object,
                    name: str, fields: dict[str, 'UFieldDeclaration'], ctx: 'ValidateContext') -> list['ValidationFailure']:
    from ...internal.validation.GetTypeUnexpectedValidationFailure import get_type_unexpected_validation_failure
    from ...internal.validation.ValidateStructFields import validate_struct_fields

    if isinstance(value, dict):
        selected_fields = cast(
            list[str], ctx.select.get(name) if ctx.select else None)
        return validate_struct_fields(fields, selected_fields, value, ctx)
    else:
        return get_type_unexpected_validation_failure([], value, _STRUCT_NAME)
